File: core/src/main/java/net/wiredtomato/waygl/core/service/PlatformService.java
package net.wiredtomato.waygl.core.service;

import java.nio.file.Path;
import java.util.ServiceLoader;

public interface PlatformService {
    PlatformService IMPL = ServiceLoader.load(PlatformService.class).findFirst().orElseThrow(() -> new IllegalStateException("Failed to load service: " + PlatformService.class));

    String platform();
    Boolean isModLoaded(String modId);
    Boolean isDevelopmentEnvironment();
    default String getEnvironmentName() {
        if (isDevelopmentEnvironment()) {
            return "development";
        } else return "production";
    }

    String getMinecraftVersionString();
    Path getConfigDirectory();
}

File: mod/common/src/main/kotlin/net/wiredtomato/waygl/service/PlatformService.kt
package net.wiredtomato.waygl.service

import net.wiredtomato.waygl.util.Version
import net.wiredtomato.waygl.util.Version.Companion.toVersion
import java.nio.file.Path

interface PlatformService {
    fun platform(): String
    fun isModLoaded(modId: String): Boolean
    fun isDevelopmentEnvironment(): Boolean
    fun getEnvironmentName(): String = if (isDevelopmentEnvironment()) "development" else "production"
    fun getMinecraftVersionString(): String
    fun getMinecraftVersion(): Version = getMinecraftVersionString().toVersion()
    fun getConfigDir(): Path
}

val PlatformServiceImpl = Services.getService<PlatformService>()
